File: tech-start-test-angular/src/app/model/Invoice.ts
import { deserialize, deserializeAs} from "cerialize";
import { Customer } from "./Customer";
import { Distributor } from "./Distibutor";

export class Invoice{
    @deserialize id: string;
    @deserialize invoiceNumber: number;
    @deserialize purchaseDate: Date;
    @deserialize totalPurchases: number;
    @deserializeAs(Customer, 'customerLocation') customerLocation: Customer;
    @deserializeAs(Distributor, 'distributorLocation') distributorLocation: Distributor;

    constructor(id : string, invoiceNumber : number, purchaseDate : Date, totalPurchases: number, customerLocation: Customer, distributorLocation: Distributor) {
        this.id = id;
        this.invoiceNumber = invoiceNumber;
        this.purchaseDate = purchaseDate;
        this.totalPurchases = totalPurchases;
        this.customerLocation = customerLocation;
        this.distributorLocation = distributorLocation;
      }
}

File: tech-start-test-api/src/main/java/com/test/techstarttestapi/model/Invoice.java
package com.test.techstarttestapi.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.springframework.data.annotation.Id;

import java.util.Date;

@NoArgsConstructor
@AllArgsConstructor
@Data
@ToString
public class Invoice {

    @Id
    private String id;
    private Integer invoiceNumber;
    @JsonFormat(pattern="yyyy-MM-dd")
    private Date purchaseDate;
    private Integer totalPurchases;
    private CustomerLocation CustomerLocation;
    private DistributorLocation DistributorLocation;

    public void addTotalPurchases(){
        this.totalPurchases = getTotalPurchases() + 1;
    }
}
